avahi-daemon/static-services.c

```diff
@@ -420,6 +449,121 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
     return;
 }
 
+static uint8_t hex(char c)
+{
+  if ((c >= '0') && (c <= '9'))
+    return c - '0';
+  if ((c >= 'A') && (c <= 'F'))
+    return c - 'A' + 10;
+  if ((c >= 'a') && (c <= 'f'))
+    return c - 'a' + 10;
+  return 0xFF;
+}
+
+static int decode_hex_buf(struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
+    uint8_t *raw_buf;
+    size_t iter;
+    size_t raw_buf_len;
+
+    if (buf_len % 2) {
+        avahi_log_error("%s: parse failure: hex value of the txt record should have an even length", u->group->filename);
+        u->failed = 1;
+        return -1;
+    }
+    raw_buf_len = buf_len / 2;
+    raw_buf = avahi_malloc(raw_buf_len);
+    for (iter = 0; iter < raw_buf_len; iter++) {
+        uint8_t high_nibble = hex(u->buf[iter * 2]);
+        uint8_t low_nibble = hex(u->buf[iter * 2 + 1]);
+
+        if (high_nibble > 0xF || low_nibble > 0xF) {
+            avahi_log_error("%s: parse failure: failed to parse hex data: invalid hex data", u->group->filename);
+            u->failed = 1;
+            avahi_free(raw_buf);
+            return -1;
+        }
+        raw_buf[iter] = (high_nibble << 4) | low_nibble;
+    }
+    *out_buf = raw_buf;
+    *out_buf_len = raw_buf_len;
+    return 0;
+}
+
+static uint8_t base64(char c)
+{
+  if (c >= 'A' && c <= 'Z')
+    return c - 'A';
+  if (c >= 'a' && c <= 'z')
+    return c - 'a' + 26;
+  if (c >= '0' && c <= '9')
+    return c - '0' + 52;
+  if (c == '+')
+    return 62;
+  if (c == '/')
+    return 63;
+  return 255;
+}
+
+static int base64_error(struct xml_userdata *u, uint8_t *raw_buf) {
+    avahi_log_error("%s: parse failure: failed to parse base64 data: invalid base64 data", u->group->filename);
+    u->failed = 1;
+    avahi_free(raw_buf);
+    return -1;
+}
+
+static int decode_base64_buf(struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
```

Comment:
By chance, I tried this record and got a crash here.  Seems it doesn't handle the value being empty which is easy when the base64 value ends in =!

`<txt-record value-format="binary-base64">MDEyMwo=</txt-record>
`

```
(gdb) bt
#0  strlen () at ../sysdeps/x86_64/strlen.S:106
#1  0x0000555555560d6d in decode_base64_buf (u=0x7fffffffe860, out_buf=0x7fffffffe578, out_buf_len=0x7fffffffe580) at static-services.c:516
#2  0x0000555555561378 in xml_end (data=0x7fffffffe860, el=0x5555557874b0 "txt-record") at static-services.c:641
#3  0x00007ffff756fb3c in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#4  0x00007ffff757045c in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#5  0x00007ffff7571d5e in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#6  0x00007ffff7572495 in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#7  0x00007ffff757478f in XML_ParseBuffer () from /lib/x86_64-linux-gnu/libexpat.so.1
#8  0x0000555555561c35 in static_service_group_load (g=0x55555577fb50) at static-services.c:828
#9  0x0000555555561d89 in load_file (n=0x55555577fb20 "/services/example.service") at static-services.c:862
#10 0x0000555555561fd4 in static_service_load (in_chroot=1) at static-services.c:919
#11 0x000055555555ceea in run_server (c=0x55555577a200 <config>) at main.c:1235
#12 0x000055555555dc7a in main (argc=1, argv=0x7fffffffeba8) at main.c:1674

```